source/extensions/quic_listeners/quiche/platform/quic_port_utils_impl.cc

```diff
@@ -0,0 +1,53 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_port_utils_impl.h"
+
+#include "envoy/network/address.h"
+
+#include "common/common/assert.h"
+#include "common/network/utility.h"
+
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+namespace quic {
+
+int QuicPickUnusedPortOrDieImpl() {
+  // Only try to get a port for limited times.
+  std::vector<Envoy::Network::Address::IpVersion> supported_versions =
+      Envoy::TestEnvironment::getIpVersionsForTest();
+  for (size_t i = 0; i < 30000; ++i) {
+    uint32_t port = 0;
+    bool available = true;
+    for (auto ip_version : supported_versions) {
```

Comment:
> but I'm not sure if listening on 127.0.0.1:444 and then ::444 would fail on the platforms we support.
What do you mean?

The returned port of this function should be able to bind to both an v4 and v6 address (in separate test) if test supports both address families. So here we need check if we can create a socket and bind it to an address in both v4 and v6.

Reply:
I replaced the version loop with checking on single version according to our discussion. Thanks for the suggestion!